lasagne/layers/merge.py

```diff
@@ -326,9 +323,36 @@ def __init__(self, incomings, merge_function, cropping=None, **kwargs):
         super(ElemwiseMergeLayer, self).__init__(incomings, **kwargs)
         self.merge_function = merge_function
         self.cropping = cropping
+        self.broadcastable = None
 
     def get_output_shape_for(self, input_shapes):
         input_shapes = autocrop_array_shapes(input_shapes, self.cropping)
+
+        input_dims = [len(shp) for shp in input_shapes]
+        if not all(input_dim == input_dims[0] for input_dim in input_dims):
+            raise ValueError('Input dimensions must be the same but were %s' %
+                             ", ".join(map(str, input_shapes)))
+
+        def broadcasting(input_dim):
+            # Identify dimensions that will be broadcasted.
+            sorted_dim = sorted(input_dim,
+                                key=lambda x: x if x is not None else None)
+            if isinstance(sorted_dim[-1], int) and sorted_dim[-1] != 1 \
+                    and all([d == 1 for d in sorted_dim[:-1]]):
+                size_after_broadcast = sorted_dim[-1]
+                broadcast = [True if d == 1 else None for d in input_dim]
+                return ((size_after_broadcast,)*len(input_dim), broadcast)
+            else:
+                return (input_dim, [None]*len(input_dim))
+
+        # if the dimension is broadcastable we replace 1's with the size
+        # after broadcasting.
+        input_dims, broadcastable = list(zip(
+            *[broadcasting(input_dim)for input_dim in zip(*input_shapes)]))
+
+        self.broadcastable = list(zip(*broadcastable))
```

Comment:
I think this was from the previous PR. It made sense to me as well, to keep this in `__init__`, since you hadn't made any comments there, I left it to be the way it was